### ISSUE TITLE
'vault ssh' in CA mode doesn't allow to select extensions to include in the certificate request

### ISSUE THREAD

**Issue by christophetd** [hashicorp/vault#5659]
Here's a role I created in the SSH secrets engine:

```
$ vault write -f ssh/roles/myrole - <<EOF                                                                                                                   
{
  "allow_user_certificates": true,
  "allowed_users": "christophetd",
  "default_extensions": [{ "permit-pty": "" }],
  "allowed_extensions": "permit-pty",
  "default_critical_options": [{ "source-address": "10.0.0.0/24" }],
  "key_type": "ca",
  "ttl": "10m"
}
EOF
```

I'm including the `permit-pty` extension by default, but don't want users to be able to have [other extensions](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys#L280-L304) (such as `permit-port-forwarding`).

This setup works very well manually:

```
$ vault write -field signed_key ssh/sign/role public_key=@.ssh/id_rsa.pub valid_principals=christophetd > signed_key.pub
$ ssh -i signed_key.pub christophetd@10.0.0.31
(SSH connection succeeds)
```

But doesn't with the `vault ssh` command:

```
$ vault ssh -mode=ca -role=role christophetd@10.0.0.31                                                                                                      
failed to sign public key /home/christophetd/.ssh/id_rsa.pub: Error making API request.

URL: PUT http://0.0.0.0:1234/v1/ssh/sign/role
Code: 400. Errors:

* extensions [permit-user-rc permit-X11-forwarding permit-agent-forwarding permit-port-forwarding] are not on allowed list

```

I'm guessing this is due to `vault ssh` trying to have a certificate signed with all the extensions? Any way to customize it?

**Comment by christophetd** [comment]
Looking at the source, it seems that it's actually a known issue/todo:

https://github.com/hashicorp/vault/blob/ebc733f4ca5d362fdfb302ac75953228585c54a2/command/ssh.go#L387-L395
```
// TODO: let the user configure these. In the interim, if users want to
// customize these values, they can produce the key themselves.
"extensions": map[string]string{
    "permit-X11-forwarding":   "",
    "permit-agent-forwarding": "",
    "permit-port-forwarding":  "",
    "permit-pty":              "",
    "permit-user-rc":          "",
}
```

I'll let a maintainer close the issue or keep it open to track this. Might send out a PR in the next few weeks

**Comment by guilhermepsantos** [comment]
Related to this:
1. It seems to me that we can't *restrict* the values of an extension (or critical option); and
2. We also can't limit `critical_options` or `extensions` to none.

### 1. Restricting Values
For example, per the [https://www.vaultproject.io/api/secret/ssh/index.html#allowed_critical_options](documentation), if I want a role to sign keys that are usable only from a given subnet, this would be the closest thing I can do:

```hcl
{
    "allowed_critical_options": "source-address",
    "default_critical_options": [
        {
            "source-address": "X.X.X.X/X"
        }
    ]
}
```

While I haven't yet tested this, nothing seems to prevent a user from specifying his or her own `critical_options` parameter with a different `source-address`, since `source-address` is an "allowed" critical options and `X.X.X.X/X` is only a *default* value.

### 2. Restricting Options
The documentation says (emphasis added and typo italicized):
> allowed_critical_options (string: "") – Specifies a comma-separated list of critical options that certificates can have when signed. **To allow any critical options, set this to an empty string. Will default to allowing any critical options.**
> 
> allowed_extensions (string: "") – Specifies a comma-separated list of extensions that certificates can have when signed. **To allow any _critical options_, set this to an empty string. Will default to allowing any extensions.** For the list of extensions, take a look at the sshd manual's AUTHORIZED_KEYS FILE FORMAT section. You should add a permit- before the name of extension to allow it.

**Comment by rfc1036** [comment]
I agree that being able to specify mandatory extensions on a key is an important feature, because it would allow to have a role which could generate keys only valid when connecting from a specific server, hence totally useless even if stolen.

**Comment by centromere** [comment]
Has this issue been backlogged? I am interested in having more fine-grained control over which extensions are requested by the client.

**Comment by pjbgf** [comment]
I have a similar use case, in which I would like to restrict the use of the critical option `force-command` to a specific command and have that enforced by Vault. 

Is this in the backlog? If not, would the HashiCorp folks be open to a PR on the subject?

**Comment by candlerb** [comment]
> It seems to me that we can't restrict the values of an extension (or critical option)

This is true, but I think that's separate to the issue originally raised.

The original issue is about the `vault ssh` _client_ having a hard-coded list of extensions that it requests.  It means that `vault ssh` is not usable when the allowed set of extensions in the ssh role is less than this.

I think it's a separate issue about how the _server_ policy can (a) disallow all extensions or options, (b) require certain extensions or critical options, (c) allow only certain values for critical options.

**Comment by keslerm** [comment]
Hit this as well, would be much nicer if the default was maybe as restrictive as the what the tutorials tend to provide. But even then a flag to override would be extremely helpful.

**Comment by candlerb** [comment]
I think the change to address the specific problem raised by this ticket should be:

1. By default, `vault ssh -mode ca` requests *no* extensions.
    * The certificate will then get the [default_extensions](https://www.vaultproject.io/api-docs/secret/ssh#default_extensions) in the signing role (if any)
2. Add a command-line flag to select a set of extensions to request, similar to `-valid-principals`

**Comment by Mwoodin123** [comment]
Yes and it’s ruining my life
